pkg/kn/commands/plugin/plugin_verifier.go

```diff
@@ -0,0 +1,252 @@
+// Copyright © 2018 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"syscall"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+// pluginVerifier verifies that existing kn commands are not overriden
+type pluginVerifier struct {
+	root        *cobra.Command
+	seenPlugins map[string]string
+}
+
+// collect errors and warnings on the way
+type errorsAndWarnings struct {
+	errors   []string
+	warnings []string
+}
+
+// Create new verifier
+func newPluginVerifier(root *cobra.Command) *pluginVerifier {
+	return &pluginVerifier{
+		root:        root,
+		seenPlugins: make(map[string]string),
+	}
+}
+
+// permission bits for execute
+const (
+	UserExecute  = 1 << 6
+	GroupExecute = 1 << 3
+	OtherExecute = 1 << 0
+)
+
+// Verify implements pathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kn command path, or a previously seen plugin.
+func (v *pluginVerifier) verify(eaw errorsAndWarnings, path string) errorsAndWarnings {
+	if v.root == nil {
+		return eaw.addError("unable to verify path with nil root")
+	}
+
+	// Verify that plugin actually exists
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		if err == os.ErrNotExist {
+			return eaw.addError("cannot find plugin in %s", path)
+		}
+		return eaw.addError("cannot stat %s: %v", path, err)
+	}
+
+	eaw = v.addErrorIfWrongPrefix(eaw, path)
+	eaw = v.addWarningIfNotExecutable(eaw, path, fileInfo)
+	eaw = v.addWarningIfAlreadySeen(eaw, path)
+	eaw = v.addErrorIfOverwritingExistingCommand(eaw, path)
+
+	return eaw
+}
+
+func (v *pluginVerifier) addWarningIfAlreadySeen(eaw errorsAndWarnings, path string) errorsAndWarnings {
+	fileName := filepath.Base(path)
+	if existingPath, ok := v.seenPlugins[fileName]; ok {
+		return eaw.addWarning("%s is ignored because it is shadowed by a equally named plugin: %s.", path, existingPath)
+	}
+	v.seenPlugins[fileName] = path
```

Comment:
Can we move the part where we check off we've seen it to the end of `verify`, so the `addWarningIf...` methods don't mutate anything but the `eaw` and not the acceptor?